### DISCUSSION TITLE
Function rgb is missing argument $green.

### DISCUSSION THREAD

**Discussion by murtazasultani** [tailwindlabs/tailwindcss#6738] [Help]
I got this error when I switched to TailwindCSS v3

package.json:
```
{
  "devDependencies": {
    "autoprefixer": "^10.4.0",
    "postcss": "^8.4.5",
    "postcss-import": "^14.0.2",
    "tailwindcss": "^3.0.7"
  },
  "dependencies": {
    "cssnano": "^5.0.14"
  },
  "scripts": {
    "dev": "bundle exec jekyll serve --watch",
    "watch": "jekyll serve --watch",
    "build": "JEKYLL_ENV=production bundle exec jekyll build"
  }
}

```

tailwind.config.js
```
module.exports = {
  mode: 'jit',
  content: [
    './_includes/**/*.html',
    './_layouts/**/*.html',
    './*.html',
  ],
  theme: {
    extend: {},
  },
  plugins: [],
}

```

postcss.config.js

```
module.exports = {
    plugins: [
      require('tailwindcss')('./tailwind.config.js'),
      require('postcss-import'),
      require('autoprefixer'),
      ...(process.env.JEKYLL_ENV == "production"
        ? [require('cssnano')({})]
        : [])
    ]
}
```

**Comment by jozefvaclavik** [comment]
Same for me. Happens with Rails (jsbundling-rails with esbuild), but only when precompiling assets for production release. Development works fine.

```
SassC::SyntaxError: Error: Function rgb is missing argument $green.
  | on line 427 of stdin
  | >>   border-color: rgb(229 231 235 / var(--tw-border-opacity));
  |  
  | ----------------^
```

**Reply by adamwathan** [reply]
It looks like for some reason your Tailwind compiled CSS is being fed into Sass and Sass doesn’t understand modern CSS color syntax (commaless rgb).

I would try to diagnose why that CSS is being fed into Sass, you definitely don’t want to do that.

**Reply by jozefvaclavik** [reply]
Ok, so by default rails (6) comes with `gem 'sass-rails'` and `app/assets/stylesheets/application.css.scss`. In my case this file had only `*= require_self`. The rabbit hole took me down to `sassc-rails`, which depends on `sassc`, which depends on LibSass, which tends to be behind with features and everyone moved to Dart.

I actually don't need sass as in my case everything is being handled through `jsbundling-rails` with `esbuild` that uses `sassPlugin` and `postcss`. After I dropped it from gemfile and removed `application.css.scss` doing `rake assets:precompile` passed.

Other people who can't drop sass-rails may not be so fortunate.

Thx @adamwathan for pointing me the right way.

**Reply by gambala** [reply]
Got same error in Ruby on Rails on sprockets precompilation step after tailwind build file generated

In Rails, it can be fixed by adding

```
config.assets.css_compressor = nil
```

to `config/environments/production.rb` file. It prevents sprockets of using sass mode and SassC gem (which based on deprecated LibSass library) in assets:precompile step

**Reply by jwilsjustin** [reply]
I made a gem that uses esbuild for CSS compression in your Rails app: https://github.com/jwilsjustin/escompress. If you are already using esbuild in your stack (but also need sass-rails for some reason) then I hope this gem is helpful.

**Reply by NoahAncolie** [reply]
@gambala Yes ! Worked !!!

**Reply by sunwooz** [reply]
Worked like a gem(finally!)

**Reply by lelouvincx** [reply]
Thanks a lot! It works

**Reply by stevehill1981** [reply]
This works beautifully, thanks @gambala!

**Reply by ElvisLYC** [reply]
Thanks, this is life-saver

**Reply by xiaohui-zhangxh** [reply]
thanks, finally get away from sass

**Reply by ugifractal** [reply]
wow, thanks

**Comment by kazuooooo** [comment]
In my case, there is a `output.css` file of tailwind build in my webpack build target.
(It means I tried build Tailwind 2 times)
So I can solve the problem by removing the output.css.

**Reply by kritik** [reply]
Monkeypatch if you don't want to switch off asset-pipeline.
```
require 'sprockets/sass_compressor'
class Sprockets::SassCompressor
  TAILWIND_SEARCH = "--tw-".freeze
  def call(*args)
    input = if defined?(data)
              data # sprockets 2.x
            else
              args[0][:data] #sprockets 3.x
            end

    return input if skip_compiling?(input) # added this line

    SassC::Engine.new(
      input,
      {
        style: :compressed
      }
    ).render
  end

  def skip_compiling?(body)
    body.include?(TAILWIND_SEARCH)
  end

end
```

**Reply by liqites** [reply]
you made my day!

**Reply by benjaminwood** [reply]
One more monkeypatch solution:

```ruby
module SassCompressorMonkeypatch
  def call_processor(processor, input)
    super
  rescue SassC::SyntaxError => e
    raise unless processor == Sprockets::SassCompressor
    puts "Warning: Could not compress #{input[:filename]} with Sprockets::SassCompressor. Returning uncompressed result."
    metadata = (input[:metadata] || {}).dup
    metadata[:data] = input[:data]
    metadata
  end
end

Sprockets::Base.prepend(SassCompressorMonkeypatch)
```

Instead of looking in the source for tailwind classes, this one simply rescues `SassC::SyntaxError` and returns the input, which in the case of tailwind is already compressed.

**Reply by Micka33** [reply]
@benjaminwood I use Tailwind and ActiveAdmin.
```ruby
gem 'activeadmin', '~> 3.2' # <- Needs sassc present in `gem "sass-rails", "~> 6.0"`
gem 'tailwindcss-rails' # <- incompatible with sassc
```

I put @benjaminwood 's  `SassCompressorMonkeypatch` in an initializer, and it works perfectly.

Thanks!

**Comment by MinRam** [comment]
rgba( R, G, B, A)

**Comment by marckohlbrugge** [comment]
Did anyone ever find a reliable solution for this?

**Reply by kritik** [reply]
@marckohlbrugge check my monkey patch

**Reply by stefanblazevski** [reply]
@marckohlbrugge If you want to or have to use the new syntax, you can use upper case RGB instead, and sass will ignore it.
This is especially useful when you are using CSS variables you don't control:
`.header{
    background-color: RGB(var(--color-blue) / 49%);
}`

**Reply by guschiavon** [reply]
@stefanblazevski can you explain how to make RGB uppercase in all instances of Tailwind?

**Comment by emcoding** [comment]
I tried the gem mentioned in the discussion, I tried different env settings for assets in `test.rb` , but it fails with `rails test:prepare` and with `RAILS_ENV=test bin/rails assets:precompile`. BUT it works without problems when I precompile in dev env:

```
% bin/rails assets:clobber
% bin/rails assets:precompile
```
 
Beats me.

**Comment by pedryvo** [comment]
On Fly.io deploys, change this line in Dockerfile:

`RUN RAILS_ENV=development SECRET_KEY_BASE=1 ./bin/rails assets:precompile`

**Reply by jurgens** [reply]
this one works. for some reason assets compilation works fine in development but not in test environment

**Reply by ksaitor** [reply]
this worked!
